### PR TITLE
Support UPF Paging request#7220

### PR DESCRIPTION
**Pull request by prabinakpattnaik**
Signed-off-by: prabina pattnaik <prabinak@wavelabs.ai>

## Summary

UPF Paging request supported in SMF:
Received the message from UPF
Get the IMSI from mobilityd w.r.t. UE IP Address.
Read the SessionMap from global session store.
Send the notification to AMF for session info.

### CODE REVIEW

File: lte/gateway/c/session_manager/CMakeLists.txt [hunk 1]
```diff
@@ -54,7 +55,7 @@ endif ()
 
 include_directories("${OUTPUT_DIR}")
 include_directories("${MAGMA_ROOT}/orc8r/gateway/c/common/logging")
-
+include_directories("${MAGMA_ROOT}/lte/gateway/c/oai/lib/mobility_client")
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/CMakeLists.txt]: We definitely do not want a dependency into `lte/gateway/c/oai/lib/mobility_client`. Let's create a mobilityd client in SessionD if we need to talk to mobilityd

**Review by themarwhal — CHANGES_REQUESTED** [review]

File: lte/protos/session_manager.proto [hunk 2]
```diff
@@ -1013,7 +1030,7 @@ message UPFPagingInfo {
 service SetInterfaceForUserPlane {
     rpc SetUPFNodeState(UPFNodeState) returns (SmContextVoid) {}
     rpc SetUPFSessionsConfig(UPFSessionConfigState) returns (magma.orc8r.Void) {}
-    rpc SendPagingReuest(UPFPagingInfo) returns (magma.orc8r.Void) {}
+    rpc SendPagingRequest(UPFPagingInfo) returns (SmContextVoid) {}
```
Comment (themarwhal) [inline lte/protos/session_manager.proto:1086]: Can we pull out this spelling mistake fix / void type change into a separate PR? Easier to approve simple changes like that.

File: lte/gateway/c/session_manager/CMakeLists.txt (hunk 1 above)
Comment (prabinakpattnaik) [reply lte/gateway/c/session_manager/CMakeLists.txt]: Yes Marie, I already fixed same things in magma-5G. I will add here.

File: lte/protos/session_manager.proto (hunk 2 above)
Comment (prabinakpattnaik) [reply lte/protos/session_manager.proto:1086]: I fixed it here.

File: lte/gateway/c/session_manager/CMakeLists.txt (hunk 1 above)
Comment (prabinakpattnaik) [reply lte/gateway/c/session_manager/CMakeLists.txt]: Changes done.

File: lte/gateway/c/session_manager/UpfMsgManageHandler.cpp [hunk 3]
```diff
@@ -123,4 +130,66 @@ void UpfMsgManageHandler::SetUPFSessionsConfig(
...
+  mobilityd_client_->get_subscriberid_from_ipv4(
+      req, [this, fte_id, response_callback](
+               Status status, const SubscriberID& sid) {
+        if (!status.ok()) {
+          MLOG(MERROR) << "Subscriber could not be found for ip ";
+        }
+        const std::string& imsi = sid.id();
+        conv_enforcer_->get_event_base().runInEventBaseThread(
+            [this, imsi, fte_id, response_callback]() {
+              if (!imsi.length()) {
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/UpfMsgManageHandler.cpp]: Can we pull out the inside of this lambda function passed into `conv_enforcer_->get_event_base().runInEventBaseThread(` into a separate function? I think it'll be cleaner to read that way.

File: lte/gateway/c/session_manager/UpfMsgManageHandler.h [hunk 4]
```diff
@@ -52,6 +53,14 @@ class UpfMsgHandler {
   virtual void SetUPFSessionsConfig(
       ServerContext* context, const UPFSessionConfigState* sess_config,
       std::function<void(Status, SmContextVoid)> response_callback) = 0;
+
+  /**
+   * Paging Notification handling
+   *
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/UpfMsgManageHandler.h]: extra line in comment

File: lte/gateway/c/session_manager/UpfMsgManageHandler.cpp [hunk 5]
```diff
@@ -123,4 +130,66 @@ void UpfMsgManageHandler::SetUPFSessionsConfig(
...
+                MLOG(MERROR) << "Subscriber is NULL";
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/UpfMsgManageHandler.cpp]: Let's be more descriptive here. 
Maybe something like
`get_subscriberid_from_ipv4 for IP <IP> returned an empty subscriber ID`

File: lte/gateway/c/session_manager/UpfMsgManageHandler.cpp [hunk 6]
```diff
@@ -123,4 +130,66 @@ void UpfMsgManageHandler::SetUPFSessionsConfig(
...
+                Status status(grpc::NOT_FOUND, "Sesion Not found");
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/UpfMsgManageHandler.cpp]: `Session` typo

Comment (themarwhal) [reply lte/gateway/c/session_manager/UpfMsgManageHandler.cpp]: but i think this error message can also be more descriptive. Specify the reason why it's failing.

File: lte/gateway/c/session_manager/UpfMsgManageHandler.cpp [hunk 7]
```diff
@@ -123,4 +130,66 @@ void UpfMsgManageHandler::SetUPFSessionsConfig(
...
+               * find  respective sesion to operate
+               */
+              SessionSearchCriteria criteria(imsi, IMSI_AND_TEID, fte_id);
+
+              auto session_it =
+                  session_store_.find_session(session_map, criteria);
+              if (!session_it) {
+                MLOG(MERROR) << "No session found in SessionMap for IMSI "
+                             << imsi << " with teid " << fte_id;
+                Status status(grpc::NOT_FOUND, "Sesion Not found");
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/UpfMsgManageHandler.cpp]: Sesion typo -> `session`

Specify in the message that the session was not found for IMSI/Teid pair.

File: lte/gateway/c/session_manager/UpfMsgManageHandler.cpp [hunk 8]
```diff
@@ -123,4 +130,66 @@ void UpfMsgManageHandler::SetUPFSessionsConfig(
...
+                response_callback(status, SmContextVoid());
+                return;
+              }
+
+              MLOG(MINFO) << "IDLE_MODE::: Session found in SendingPaging "
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/UpfMsgManageHandler.cpp]: Log the session ID here.

**Review by themarwhal — CHANGES_REQUESTED** [review]
Change in general looks good. Left some comments.


Thank you Marie for reviewed the code. Addressed the review comments.

File: lte/gateway/c/session_manager/UpfMsgManageHandler.cpp (hunk 8 above)
Comment (prabinakpattnaik) [reply lte/gateway/c/session_manager/UpfMsgManageHandler.cpp]: Added session_id in a log statement.

File: lte/gateway/c/session_manager/UpfMsgManageHandler.cpp (hunk 7 above)
Comment (prabinakpattnaik) [reply lte/gateway/c/session_manager/UpfMsgManageHandler.cpp]: Yes. Corrected typo and changed status message.

File: lte/gateway/c/session_manager/UpfMsgManageHandler.cpp (hunk 6 above)
Comment (prabinakpattnaik) [reply lte/gateway/c/session_manager/UpfMsgManageHandler.cpp]: Fixed.

File: lte/gateway/c/session_manager/UpfMsgManageHandler.cpp (hunk 5 above)
Comment (prabinakpattnaik) [reply lte/gateway/c/session_manager/UpfMsgManageHandler.cpp]: Changed the error log statement.

File: lte/gateway/c/session_manager/UpfMsgManageHandler.h (hunk 4 above)
Comment (prabinakpattnaik) [reply lte/gateway/c/session_manager/UpfMsgManageHandler.h]: Fixed.

File: lte/gateway/c/session_manager/UpfMsgManageHandler.cpp (hunk 3 above)
Comment (prabinakpattnaik) [reply lte/gateway/c/session_manager/UpfMsgManageHandler.cpp]: Created a new function and lambad function is defined inside a new function.

File: lte/gateway/c/session_manager/MobilitydClient.cpp [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+#include "MobilitydClient.h"
+#include "includes/ServiceRegistrySingleton.h"
+#include "magma_logging.h"
+
+using grpc::Status;
+
+namespace magma {
+
+AsyncMobilitydClient::AsyncMobilitydClient(
+    std::shared_ptr<grpc::Channel> channel)
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/MobilitydClient.cpp:25]: pls address
```
Add #include <memory> for shared_ptr<>  [build/include_what_you_use] [4]
```

File: lte/gateway/c/session_manager/MobilitydClient.cpp [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+AsyncMobilitydClient::AsyncMobilitydClient()
+    : AsyncMobilitydClient(ServiceRegistrySingleton::Instance()->GetGrpcChannel(
+          "mobilityd", ServiceRegistrySingleton::LOCAL)) {}
+
+void AsyncMobilitydClient::get_subscriberid_from_ipv4(
+    const IPAddress& ue_ip_addr,
+    std::function<void(Status status, SubscriberID)> callback) {
+  auto local_resp = new AsyncLocalResponse<SubscriberID>(
+      std::move(callback), RESPONSE_TIMEOUT);
+  local_resp->set_response_reader(std::move(stub_->AsyncGetSubscriberIDFromIP(
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/MobilitydClient.cpp:37]: pls address
```
 Add #include <utility> for move  [build/include_what_you_use] [4]
```

File: lte/gateway/c/session_manager/MobilitydClient.h [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+/**
+ * AsyncMobilitydClient sends asynchronous calls to Mobilityd to retrieve
+ * UE information.
+ */
+class AsyncMobilitydClient : public GRPCReceiver, public MobilitydClient {
+ public:
+  AsyncMobilitydClient();
+
+  AsyncMobilitydClient(std::shared_ptr<grpc::Channel> mobilityd_channel);
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/MobilitydClient.h]:
```suggestion
  explicit AsyncMobilitydClient(std::shared_ptr<grpc::Channel> mobilityd_channel);
```

File: lte/gateway/c/session_manager/MobilitydClient.h [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+  /**
+   * Get SubscriberID for correspoding of UE_IP
+   */
+  void get_subscriberid_from_ipv4(
+      const IPAddress& ue_ip_addr,
+      std::function<void(Status status, SubscriberID)> callback);
+
+ private:
+  static const uint32_t RESPONSE_TIMEOUT = 6;  // seconds
+  std::unique_ptr<MobilityService::Stub> stub_;
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/MobilitydClient.h:64]: pls address
```
 Add #include <memory> for unique_ptr<>  [build/include_what_you_use] [4]
```

File: lte/gateway/c/session_manager/UpfMsgManageHandler.cpp [hunk 13]
```diff
@@ -123,4 +130,75 @@ void UpfMsgManageHandler::SetUPFSessionsConfig(
...
+                      "returned an empty subscriber ID";
+      Status status(
+          grpc::NOT_FOUND,
+          "Session Not found because"
+          "subscriber ID not found for IP");
+      response_callback(status, SmContextVoid());
+      return;
+    }
+
+    // retreive session_map entry
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/UpfMsgManageHandler.cpp]: please address 
```
[misspell] reported by reviewdog 🐶
"retreive" is a misspelling of "retrieve"
```

File: lte/gateway/c/session_manager/UpfMsgManageHandler.cpp [hunk 14]
```diff
@@ -123,4 +130,75 @@ void UpfMsgManageHandler::SetUPFSessionsConfig(
...
+    }
+
+    auto& session = **session_it;
+    MLOG(MINFO) << "IDLE_MODE::: Session found in SendingPaging "
+                   "Request of imsi: "
+                << imsi << "  session_id: " << session->get_session_id();
+    // Generate Paging trigget to AMF.
+    conv_enforcer_->handle_state_update_to_amf(
+        *session, magma::lte::M5GSMCause::OPERATION_SUCCESS, UE_PAGING_NOTIFY);
+    MLOG(MINFO) << "UPF Paging notificaiton forwarded to AMF of imsi:" << imsi;
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/UpfMsgManageHandler.cpp]: pls address
```
[misspell] reported by reviewdog 🐶
"notificaiton" is a misspelling of "notification"
```

**Review by themarwhal — CHANGES_REQUESTED** [review]
please address some of the lint annotations
(especially the #include ones and spelling)

**Comment by themarwhal** [comment]
@pshelar  / @koolzz do you mind reviewing the pipelined portion of this PR? (it's a one liner)

**Review by koolzz — APPROVED** [review]
pipelined oneliner looks good

File: lte/gateway/c/session_manager/UpfMsgManageHandler.cpp (hunk 14 above)
Comment (prabinakpattnaik) [reply lte/gateway/c/session_manager/UpfMsgManageHandler.cpp]: Correction done.

File: lte/gateway/c/session_manager/UpfMsgManageHandler.cpp (hunk 13 above)
Comment (prabinakpattnaik) [reply lte/gateway/c/session_manager/UpfMsgManageHandler.cpp]: correction done.

File: lte/gateway/c/session_manager/MobilitydClient.h (hunk 12 above)
Comment (prabinakpattnaik) [reply lte/gateway/c/session_manager/MobilitydClient.h:64]: Addressed it.

File: lte/gateway/c/session_manager/MobilitydClient.h (hunk 11 above)
Comment (prabinakpattnaik) [reply lte/gateway/c/session_manager/MobilitydClient.h]: changes done as suggested.

File: lte/gateway/c/session_manager/MobilitydClient.cpp (hunk 10 above)
Comment (prabinakpattnaik) [reply lte/gateway/c/session_manager/MobilitydClient.cpp:37]: Addressed it.

File: lte/gateway/c/session_manager/MobilitydClient.cpp (hunk 9 above)
Comment (prabinakpattnaik) [reply lte/gateway/c/session_manager/MobilitydClient.cpp:25]: Addressed it.

**Comment by prabinakpattnaik** [comment]
> please address some of the lint annotations
> (especially the #include ones and spelling)

Addressed review comments and rebased with latest master. Thanks Marie.

**Review by themarwhal — APPROVED** [review]

**Pull request merged**